### PR TITLE
Use contract assert for CUDA API failure

### PR DESCRIPTION
**Pull request by ibaned**
See #2173 for the reason why I created this prototype branch.

### CODE REVIEW

**Comment by ibaned** [comment]
spot-check of the prototype
```
Running on machine: kokkos-dev-2
Repository Status:  f58120a39b1e46f0b73ea922728fa27ed8700a6a Fix several bugs


Going to test compilers:  gcc/7.3.0 gcc/9.1 intel/18.0.5 clang/8.0 cuda/10.1
Testing compiler gcc/7.3.0
  Starting job gcc-7.3.0-OpenMP-release
  PASSED gcc-7.3.0-OpenMP-release
Testing compiler gcc/9.1
  Starting job gcc-7.3.0-Pthread-release
  PASSED gcc-7.3.0-Pthread-release
  Starting job gcc-9.1-OpenMP-release
  PASSED gcc-9.1-OpenMP-release
Testing compiler intel/18.0.5
  Starting job gcc-9.1-Serial-release
  PASSED gcc-9.1-Serial-release
Testing compiler clang/8.0
  Starting job intel-18.0.5-OpenMP-release
  PASSED intel-18.0.5-OpenMP-release
  Starting job clang-8.0-Cuda_OpenMP-release
  PASSED clang-8.0-Cuda_OpenMP-release
Testing compiler cuda/10.1
  Starting job clang-8.0-Pthread_Serial-release
  PASSED clang-8.0-Pthread_Serial-release
  Starting job cuda-10.1-Cuda_OpenMP-release
  PASSED cuda-10.1-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=166 run_time=145
clang-8.0-Pthread_Serial-release build_time=91 run_time=275
cuda-10.1-Cuda_OpenMP-release build_time=238 run_time=157
gcc-7.3.0-OpenMP-release build_time=94 run_time=104
gcc-7.3.0-Pthread-release build_time=84 run_time=264
gcc-9.1-OpenMP-release build_time=92 run_time=101
gcc-9.1-Serial-release build_time=74 run_time=170
intel-18.0.5-OpenMP-release build_time=348 run_time=133
```

File: core/src/impl/Kokkos_Error.hpp [hunk 1]
```diff
@@ -66,6 +76,17 @@ void traceback_callstack( std::ostream & );
 
 std::string human_memory_size(size_t arg_bytes);
```
Comment (mhoemmen) [inline core/src/impl/Kokkos_Error.hpp]: https://en.wikipedia.org/wiki/The_Magical_Number_Seven,_Plus_or_Minus_Two

File: core/src/Cuda/Kokkos_Cuda_Error.hpp [hunk 2]
```diff
@@ -51,15 +51,15 @@ namespace Kokkos { namespace Impl {
...
+void cuda_internal_error_throw( cudaError e , const char * name, const char * file, char const* function, const int line );
 
-inline void cuda_internal_safe_call( cudaError e , const char * name, const char * file = NULL, const int line = 0)
+inline void cuda_internal_safe_call( cudaError e , const char * name, const char * file, char const* function, const int line)
 {
-  if ( cudaSuccess != e ) { cuda_internal_error_throw( e , name, file, line ); }
+  if ( cudaSuccess != e ) { cuda_internal_error_throw( e , name, file, function, line ); }
 }
 
 #define CUDA_SAFE_CALL( call )  \
```
[the remark this answers is not in the document]
Comment (ibaned) [reply core/src/Cuda/Kokkos_Cuda_Error.hpp]: Agreed, just doing the minimal viable amount of work in this PR

File: core/src/impl/Kokkos_Error.hpp (hunk 1 above)
Comment (ibaned) [reply core/src/impl/Kokkos_Error.hpp]: I honestly don't know what this function does :)

**Comment by mhoemmen** [comment]
@ibaned I was going to ask whether the stack trace thing actually works or if it's just a stub for now

**Comment by ibaned** [comment]
@mhoemmen the stack trace function in Kokkos is `ifdef`'ed out by a macro that doesn't start with `KOKKOS_` and doesn't show up anywhere in our build systems, so it seems like it was only enabled manually for some developer's debugging. Also, it doesn't use `libbfd` or alternatives, so it may not give line numbers in most cases.

**Comment by mhoemmen** [comment]
@ibaned If there's a way we could write a test for this, I would be perfectly happy just merging it straight into both Kokkos and Trilinos.  I don't think stack traces have a standard format, but we could at least test that it prints something to `stderr` and returns with a nonzero error code.  We could also test whether it's possible to set a handler.

**Comment by mhoemmen** [comment]
I've been looking at Teuchos' stack tracing.  It depends on the BinUtils TPL.  If I could get a bit of help adding this to Kokkos for the non-Trilinos build system, I could get started on this.

**Comment by mhoemmen** [comment]
I wrote a prototype stack trace that only depends on POSIX functions.  [Here is the WIP PR.](https://github.com/trilinos/Trilinos/pull/5493).  I had heard that `dlopen` only works with shared object files, so I restricted myself to `backtrace` and `backtrace_symbols`, with optional demangling.

Teuchos' stack tracing has an option to set a signal handler, but Teuchos does nonreëntrant things in the signal handler, so we can't use that directly.  That's also out of scope -- we already have a non-SIGSEGV way of catching bounds errors in Kokkos, and that's the thing that we want to catch here.

**Comment by mhoemmen** [comment]
I've been looking at `cmake/kokkos_options.cmake`, and I think I understand how to add a TPL dependency.  That would let us treat `backtrace*` and the cxxabi demangle function as (two separate) TPL dependencies.  Each would only need a header file (the `backtrace*` functions need `execinfo.h`, and the cxxabi needs `cxxabi.h`).

**Pull request closed without merging**